generator/cities_boundaries_builder.cpp

```diff
@@ -0,0 +1,106 @@
+#include "generator/cities_boundaries_builder.hpp"
+
+#include "generator/utils.hpp"
+
+#include "search/categories_cache.hpp"
+#include "search/cbv.hpp"
+#include "search/mwm_context.hpp"
+
+#include "indexer/cities_boundaries_serdes.hpp"
+#include "indexer/city_boundary.hpp"
+#include "indexer/classificator.hpp"
+#include "indexer/index.hpp"
+#include "indexer/mwm_set.hpp"
+
+#include "platform/local_country_file.hpp"
+
+#include "coding/compressed_bit_vector.hpp"
+
+#include "base/assert.hpp"
+#include "base/checked_cast.hpp"
+
+#include <map>
+#include <vector>
+
+#include "defines.hpp"
+
+using namespace indexer;
+using namespace search;
+using namespace std;
+
+namespace generator
+{
+namespace
+{
+bool ParseFeatureIdToOsmIdMapping(string const & path, map<uint32_t, vector<osm::Id>> & mapping)
+{
+  return ForEachOsmId2FeatureId(path, [&](osm::Id const & osmId, uint32_t const featureId) {
+    mapping[featureId].push_back(osmId);
+  });
+}
+
+struct LocalitiesSource
+{
+  LocalitiesSource()
+  {
+    auto & c = classif();
+    m_city = c.GetTypeByPath({"place", "city"});
+    m_town = c.GetTypeByPath({"place", "town"});
+  }
+
+  template <typename Fn>
+  void ForEachType(Fn && fn) const
+  {
+    fn(m_city);
+    fn(m_town);
+  }
+
+  uint32_t m_city = 0;
+  uint32_t m_town = 0;
+};
+
+CBV GetLocalities(string const & dataPath)
+{
+  Index index;
+  auto const result = index.Register(platform::LocalCountryFile::MakeTemporary(dataPath));
+  CHECK_EQUAL(result.second, MwmSet::RegResult::Success, ("Can't register", dataPath));
+
+  search::MwmContext context(index.GetMwmHandleById(result.first));
+  return search::CategoriesCache(LocalitiesSource{}, my::Cancellable{}).Get(context);
```

Comment:
Are you sure there's no use-after-free here, when the cache is destroyed at scope exit?

Reply:
Ah, CategoriesCache::Get() returns by copy. Ok.

Reply:
Cache exists until the end of the full expression, no UAF in any case.